presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java

```diff
@@ -392,6 +409,40 @@ public void enforceQueryMaxRunTimeLimits()
     }
 
     /**
+     * Enforce max driver number at the query level
+     */
+    public void enforceQueryMaxTasksLimits()
+    {
+        for (QueryExecution query : queries.values()) {
+            if (query.getQueryInfo().getState().isDone()) {
+                continue;
+            }
+            int queryMaxDrivers = SystemSessionProperties.getQueryMaxDrivers(query.getSession());
+            int totalDrivers = query.getQueryInfo().getQueryStats().getTotalDrivers();
+            if (queryMaxDrivers < totalDrivers) {
+                query.fail(new PrestoException(EXCEEDED_DRIVER_NUMBER_LIMIT, "Query exceeded maximum driver number limit of " + queryMaxDrivers));
```

Comment:
Valid point about splits. What I've been looking for, which I'm realizing this patch doesn't address, is a way to run a check before the query starts. Although I don't think we have enough info to measure that currently. Allowing a query to run for a long period before killing it once it reaches a threshold (which I _think_ is what's happening here) is not ideal.